File: solutions/aoc2021_day07.js
var fs = require('fs');

const median = (arr) => {
    const mid = Math.floor(arr.length / 2);
    var nums = [...arr].sort((a, b) => a - b);
    return arr.length % 2 !== 0 ? nums[mid] : (nums[mid - 1] + nums[mid]) / 2;
};

function fuel_cost(positions, m, method) {
    var cost = 0;
    for (var c of positions) {
        if (method === "constant") {
            cost += Math.abs(m - c);
        }
        else { // "step"
            cost += Math.abs(m - c) * (Math.abs(m - c) + 1) / 2;
        }
    }
    return cost;
}

function min_fuel_cost(positions, method) {
    var m = median(positions);
    var cost = {};
    cost[m] = fuel_cost(positions, m, method);
    cost[m-1] = fuel_cost(positions, m-1, method);
    cost[m+1] = fuel_cost(positions, m+1, method);
    var n = (cost[m-1] < cost[m]) ? m-1 : (cost[m+1] < cost[m]) ? m+1 : m;
    if (n === m) return cost[m];
    var incr = (n === m-1) ? -1 : +1;
    var prev = m;
    while (cost[n] < cost[prev]) {
        prev = n;
        n += incr;
        cost[n] = fuel_cost(positions, n, method);
    }
    return cost[prev];
}

function part1(pi) {
    var positions = pi.split(',').map(Number);
    var ans = min_fuel_cost(positions, "constant");
    console.log(`part1: ${ans}`);
}

function part2(pi) {
    var positions = pi.split(',').map(Number);
    var ans = min_fuel_cost(positions, "step");
    console.log(`part2: ${ans}`);
}

console.log("day07");
var f = fs.readFileSync('aoc2021_day07.txt', 'utf8');
var pi = f.toString();
part1(pi);
part2(pi);

File: solutions/aoc2021_day07.py
import copy
import statistics

def fuel_cost(positions, m, method="constant"):
    cost = 0
    for c in positions:
        if method == "constant":
            cost += abs(m - c)
        else: # "step"
            cost += abs(m - c) * (abs(m - c) + 1) // 2
    return cost

def min_fuel_cost(positions, method="constant"):
    m = int(statistics.median(positions))
    cost = {m: fuel_cost(positions, m, method),
        m-1: fuel_cost(positions, m-1, method),
        m+1: fuel_cost(positions, m+1, method)}
    n = m-1 if cost[m-1] < cost[m] else m+1 if cost[m+1] < cost[m] else m
    if n == m: return cost[m]
    incr = -1 if n == m-1 else +1
    prev = m
    while cost[n] < cost[prev]:
        prev = n
        n += incr
        cost[n] = fuel_cost(positions, n, method)
    return cost[prev]

def part1(arr):
    positions = [int(e) for e in arr[0].split(',')]
    ans = min_fuel_cost(positions)
    print(ans)

def part2(arr):
    positions = [int(e) for e in arr[0].split(',')]
    ans = min_fuel_cost(positions, method="step")
    print(ans)

with open("aoc2021_day07.txt") as f:
    data = f.read()
    arr = [line for line in data.splitlines()]

print("day07")
part1(arr)
part2(arr)
